openedx/core/djangoapps/content_libraries/docs/decisions/0002-libraries-lti-provider.rst

```diff
@@ -0,0 +1,42 @@
+Allow content libraries to be used by LTI consumers
+---------------------------------------------------
+
+Status
+------
+
+Pending
+
+Context
+-------
+
+Currently, there's no way for xblocks in blockstore-based content libraries to be served to other LTI consumers. This ADR explores ways in which this can be achieved, without overly complicating the codebase. The platform currently does have an :code:`lms.lti_provider` app which implements the LTI 1.1 spec, but it only works for modulestore-based courses, and the LTI 1.1 spec itself is also deprecated and deemed insecure.
+
+Decision
+--------
+
+We will use the latest LTI spec (`LTI 1.3`_) to serve the xblocks. This would be independent of the existing LTI 1.1 implementation, and will not require any changes in that app.
+
+Instead, we will add additional code to :code:`openedx.core.content_libraries` application to handle LTI endpoints and communicate with the LTI consumers. We will use the actively-maintained `pylti1.3`_ package which can handle the LTI-spec implementation, which saves a lot of development and testing effort, and keeps the platform codebase relevant to openedx. The library is maintained by `@dmitry-viskov`_, who also proposed using it in an `earlier PR`_ to add LTI 1.3 support for courses. `pylti1.3`_ exists as an up-to-date `PyPi package`_, and is also used by projects like `my-learning-analytics`_, and `NimbleWise (source)`_.
+
+Only two aspects of the LTI 1.3 spec family will be supported: Resource link launches and Assignment and Grading Services. Additional aspects of the implementation will not be supported, in particular, `LTI NRPS`_ and Deep Linking. Although they provide value for customers, we will focus on leaving the door open for them in the future.
+
+The rest of the functionality would be similar to how the current LTI 1.1 provider works, having similar authentication, setup, and grade syncing, except for any changes required for the LTI 1.3 protocol. However, the new endpoints would be specific to blockstore-based content libraries and would not directly interact with modulestore-courses in any form. Authorization will rely on an association between the LTI platform and libraries, exclusively, and will not support public LTI launches. Site operators will manage this 1:N relationship through the admin interface.
```

Comment:
I think 1:N is a problem here, and that each library needs to be able to host content for multiple external LMS platforms. I've seen that use case several times, and not supporting it will lead to people duplicating entire libraries just to authorize additional platforms, something we definitely don't want. It should be N:N I think (a single library can serve content to multiple external LMS platforms, and/or a single external LMS platform can consume content from multiple Open edX libraries without being configured separately each time).

Reply:
Oh, yes, we [discussed this][1]:

[1]: https://github.com/edx/edx-platform/pull/27089#issuecomment-845357272

> If that sounds too open, what do you think of having a 1:N association between LTI platforms and Libraries? During the LTI launch, we can authorize based on this association. It means only a specific platform and its users can access the library's content. It would require an additional admin setup for customers though. An certainly would be great if we could add this to the UI (e.g. "Enable LTI" button per library with previously configured platform selection).
> 
> Of course, other authorization schemas are also possible. But the above seems a good tradeoff between complexity and security.

Considering there are use-cases for N:M, I think it makes sense to change your mind about it at this point. Please, just confirm if this is your final stand, and I'll update the doc, and of course, [the patch](https://github.com/edx/edx-platform/pull/27411).

Reply:
> Oh, yes, we discussed this:

Ah, sorry, I guess I didn't fully realize the implications of what you were saying at the time, nor did I reply usefully. I wish I had flagged this for you earlier, and I hope it won't require too much re-working.

I definitely think we'll need to support multiple LTI platforms using a single library. In fact a different group at MIT already has mentioned that use case to us, and it's come up with other projects in the past.

Reply:
@bradenmacdonald:

> I definitely think we'll need to support multiple LTI platforms using a single library.

Updated the ADR accordingly. I'll start working on the changes in the code to reflect this decision. Thank you.